ChainSafe/gossamer
# task(lib/grandpa): ensure messages are stored and re-processed when needed

## Changes



- We are storing grandpa notification messages in cache to make sure that we send them only one. If the message was a consensus messages, we were sending it even if it was already sent.
- With this PR, we will not be sending consensus messages again, if they were already sent. The receiver keeps track of received messages that they could not process, so that they can be processed in future. (check usage of `addVote` for more context)

## Tests


- check that tests for `dot/network` and `lib/grandpa` passes and so does other relevant tests
- run grandpa rounds, check that we don't see votes that we already have

## Issues



- #1855

## Primary Reviewer



- @noot
